### PR TITLE
Simplify link generation code

### PR DESCRIPTION
**Pull request by rynowak**
- Removes ProvidedValues and IsBound

- Removes best-effort link generation

- simplify code where possible

- lots of test simplification

### CODE REVIEW

File: test/Microsoft.AspNet.Routing.Tests/Template/TemplateRouteTest.cs [hunk 1]
```diff
@@ -1141,92 +1130,6 @@ public void GetVirtualPathWithNonParameterConstraintReturnsUrlWithoutQueryString
             target.VerifyAll();
         }
 
-        [Fact]
-        public void GetVirtualPath_Sends_ProvidedValues()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Template/TemplateRouteTest.cs]: Provided values was removed

File: test/Microsoft.AspNet.Routing.Tests/Tree/TreeRouterTest.cs [hunk 2]
```diff
@@ -1356,41 +1153,6 @@ public void TreeRouter_GenerateLink_RejectedByFirstRoute()
         }
 
         [Fact]
-        public void TreeRouter_GenerateLink_RejectedByHandler()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/Tree/TreeRouterTest.cs]: Handlers no longer have the ability to reject a link

File: src/Microsoft.AspNet.Routing.Abstractions/VirtualPathContext.cs [hunk 3]
```diff
@@ -30,14 +29,10 @@ public class VirtualPathContext
 
         public string RouteName { get; }
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Routing.Abstractions/VirtualPathContext.cs]: Could you document these values whilst you're editing this file?

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs [hunk 4]
```diff
@@ -286,306 +285,13 @@ public void GetVirtualPath_AmbiguousRoutes_RequiresRouteValueValidation_Error()
...
-            Assert.Empty(pathData.DataTokens);
-
-            // All of these should be called
-            route1.Verify(r => r.GetVirtualPath(It.IsAny<VirtualPathContext>()), Times.Once());
-            route2.Verify(r => r.GetVirtualPath(It.IsAny<VirtualPathContext>()), Times.Once());
-            route3.Verify(r => r.GetVirtualPath(It.IsAny<VirtualPathContext>()), Times.Once());
-        }
-
-        [Fact]
-        public void GetVirtualPath_NoBestEffort_NoMatch()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: Is this still covered?

**Comment by pranavkm** [comment]
:shipit:

File: src/Microsoft.AspNet.Routing.Abstractions/VirtualPathContext.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Routing.Abstractions/VirtualPathContext.cs]: Sure

File: test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs (hunk 4 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Routing.Tests/RouteCollectionTest.cs]: Yeah good point, I'll leave this one.

**Pull request closed without merging**